Add Nomnoml Renderer

rjwignar opened the pull request:
This fixes #466 

Summary of Changes
---
This adds Nomnoml rendering support to ChatCraft using [skanaar/nomnoml](https://github.com/skanaar/nomnoml).

This feature was heavily based on [Mermaid rendering support](https://github.com/tarasglek/chatcraft.org/pull/48/files) using [mermaid-js](https://github.com/mermaid-js/mermaid). For example, Nomnoml rendering is handled by [NomnomlPreview.tsx](https://github.com/tarasglek/chatcraft.org/blob/bb7c55737566be0231fdb525ec725bb8b8409680/src/components/NomnomlPreview.tsx) similar to how Mermaid rendering is handled by MermaidPreview.tsx, and similar to mermaid, the rendering is done in Markdown.tsx when a nomnoml code block is detected:
https://github.com/tarasglek/chatcraft.org/blob/bb7c55737566be0231fdb525ec725bb8b8409680/src/components/Markdown.tsx#L108-L115

Differences from Mermaid rendering
---
Although this feature is modelled off of Mermaid rendering, due to the differences between the nomnoml and mermaid-js libraries I had to implement Nomnoml rendering differently:
### No Error Message in the Nomnoml Preview
Whenever mermaidjs fails to render mermaid code, it renders an image in the preview, but
Nomnoml currently does not render an error image, so failed renders are rendered as a blank diagram instead:
![image](https://github.com/tarasglek/chatcraft.org/assets/78163326/1910007a-fb36-4a8c-9bc1-253a412b7e4b)


### No SVG ID value support
Unlike the [mermaid-js render method](https://github.com/mermaid-js/mermaid/blob/8e95c4db55c10b152e262d6fb271415f86f616d6/packages/mermaid/src/mermaid.ts#L359-L406), nomnoml's [renderSvg method](https://github.com/skanaar/nomnoml/blob/89a1ac4470f27f5f042c156ec53e7a328f27e507/src/nomnoml.ts#L60-L71) does not support an ID parameter, so in NomnomlPreview.tsx an ID value is not generated or assigned to the rendered SVG.

### No config method/no way to directly modify default styling
Nomnoml currently (as of **1.6.2**) **does not** expose a config method like [mermaid.initialize](https://mermaid.js.org/config/usage.html#using-the-mermaidapi-initialize-mermaid-initialize-call).
This means there's currently no way to programatically modify the default styling of nomnoml diagrams.
However, nomnoml has [directives](https://github.com/skanaar/nomnoml?tab=readme-ov-file#directives) that modify styling

#### Alternative Styling Modification
In the future, I see two ways users can reliably indirectly modify the styling:
- The user can update the system prompt with instructions for nomnoml styling
- Within a ChatCraft conversation, the user iteratively ask the LLM to modify the styling

However, both of these methods are heavily dependent on the LLM.
Even GPT-4 seems to [hallucinate on nomnoml syntax](https://issue-466.console-overthinker-dev.pages.dev/api/share/chatcraft_dev/AkWRUWWG5zYmNsg-Z2e9R) when asked to modify styling, unless a [directive example is provided](https://issue-466.console-overthinker-dev.pages.dev/api/share/chatcraft_dev/UW08evlCIutmdi_v1poCZ)

humphd reviewed (approved):
Code looks good to me.

We should consider an update to the default system prompt to use this.

rjwignar commented:
Hi @tarasglek, @humphd, I've updated the default system prompt following your suggestions.
I've also added new changes to the PR and updated my PR summary. I'm interested to see if you have any thoughts or suggestions on my observations when making this PR.

humphd reviewed (approved):
Love it!

<img width="932" alt="Screenshot 2024-03-26 at 7 11 19 PM" src="https://github.com/tarasglek/chatcraft.org/assets/427398/96a15a79-2ed7-4960-b6fd-fd3060e7a9d7">

tarasglek commented:
> Love it!
> 
> <img alt="Screenshot 2024-03-26 at 7 11 19 PM" width="932" src="https://private-user-images.githubusercontent.com/427398/317069181-96a15a79-2ed7-4960-b6fd-fd3060e7a9d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1MzI1NTAsIm5iZiI6MTcxMTUzMjI1MCwicGF0aCI6Ii80MjczOTgvMzE3MDY5MTgxLTk2YTE1YTc5LTJlZDctNDk2MC1iNmZkLWZkMzA2MGU3YTlkNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyN1QwOTM3MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYjk0Y2IzN2Q2NjFlNTJjNjhhMGIwMThjYzAwYzA3YjY0YmJhMDcyMGZjNGRhYTNjNjUxNjI3NTIwNDhkYWYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.yl-sLq8jlEVtCEiQiNolyB6uYVPCGgn54iabm_rDXXk">

This is fantastic diagram to use in a blog :)

rjwignar commented:
Recently updated this PR with an error message that renders in the Nomnoml preview if the nomnoml code cannot be rendered:
![image](https://github.com/tarasglek/chatcraft.org/assets/78163326/e18c42ee-5236-431d-a86f-5cbc337c3d53)

Copying the preview card still copies the actual error:
![nomnomlCopyError](https://github.com/tarasglek/chatcraft.org/assets/78163326/79797a32-5c22-4e9a-b719-de3333238814)

WangGithub0 commented on src/lib/system-prompt.ts:
```diff
@@ -8,7 +8,7 @@ I follow these rules when responding:
 - ALWAYS include the programming language name (js) or type of data (csv) at the start of Markdown code blocks
 - Format ALL lines of code to 80 characters or fewer
-- Use Mermaid diagrams when discussing visual topics
+- Use nomnoml or Mermaid diagrams when discussing visual topics
```
should you use a capital N for "Nomnoml" ?

WangGithub0 reviewed (approved):
It's so cool!  Looks good to me!
